test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,828 @@
+        tool = ToolProvider.getSystemJavaCompiler();
+        for (int i = 9; i <= Runtime.version().feature(); i++) {
+            JavacTask ct = (JavacTask) tool.getTask(null, null, null,
+                    List.of("--release", String.valueOf(i)),
+                    null,
+                    Collections.singletonList(SimpleJavaFileObject.forSource(URI.create("myfo:/Test.java"), "")));
+            ct.analyze();
+
+            String version = String.valueOf(i);
+            Elements elements = ct.getElements();
+            elements.getModuleElement("java.base"); // forces module graph to be instantiated
+            elements.getAllModuleElements().forEach(me ->
+                    processModuleElement(me, version, ct));
+        }
+    }
+
+    private void processModuleElement(ModuleElement moduleElement, String releaseVersion, JavacTask ct) {
+        for (ModuleElement.ExportsDirective ed : ElementFilter.exportsIn(moduleElement.getDirectives())) {
+            persistElement(moduleElement, moduleElement, ct.getTypes(), releaseVersion);
+            if (ed.getTargetModules() == null) {
+                processPackageElement(ed.getPackage(), releaseVersion, ct);
+            }
+        }
+    }
+
+    private void processPackageElement(PackageElement pe, String releaseVersion, JavacTask ct) {
+        persistElement(pe, pe, ct.getTypes(), releaseVersion);
+        List<TypeElement> typeElements = ElementFilter.typesIn(pe.getEnclosedElements());
+        for (TypeElement te : typeElements) {
+            processClassElement(te, releaseVersion, ct.getTypes(), ct.getElements());
+        }
+    }
+
+    /// JDK documentation only contains public and protected declarations
+    private boolean isDocumented(Element te) {
+        Set<Modifier> mod = te.getModifiers();
+        return mod.contains(Modifier.PUBLIC) || mod.contains(Modifier.PROTECTED);
+    }
+
+    private boolean isMember(Element e) {
+        var kind = e.getKind();
+        return kind.isField() || switch (kind) {
+            case METHOD, CONSTRUCTOR -> true;
+            default -> false;
+        };
+    }
+
+    private void processClassElement(TypeElement te, String version, Types types, Elements elements) {
+        if (!isDocumented(te)) {
+            return;
+        }
+        persistElement(te.getEnclosingElement(), te, types, version);
+        elements.getAllMembers(te).stream()
+                .filter(this::isDocumented)
+                .filter(this::isMember)
+                .forEach(element -> persistElement(te, element, types, version));
+        te.getEnclosedElements().stream()
+                .filter(element -> element.getKind().isDeclaredType())
+                .map(TypeElement.class::cast)
+                .forEach(nestedClass -> processClassElement(nestedClass, version, types, elements));
+    }
+
+    public void persistElement(Element explicitOwner, Element element, Types types, String version) {
```

Comment:
I'm not sure `persistElement` is a good name for this, it sounds like the element is being stored permanently (beyond the runtime of this program). I can't think of a really good name, but I think `processElement` would be ok.